wremnants/CardTool.py

```diff
@@ -715,7 +731,774 @@ def writeOutput(self, args=None, xnorm=False, forceNonzero=True, check_systs=Tru
         if self.skipHist:
             logger.info("Histograms will not be written because 'skipHist' flag is set to True")
         self.writeCard()
-        logger.info(f"Output stored in {self.outfolder}")
+
+    def writeHDF5Output(self, args=None, xnorm=False, theoryFit=False,
+        simultaneousABCD=False,
+        forceNonzero=True, check_systs=False, allowNegativeExpectation=False, 
+        sparse=False, dtype="float64", chunkSize=4*1024**2,
+        logkepsilon=math.log(1e-3), #numerical cutoff in case of zeros in systematic variations
+        clipSystVariations=False,
+        clipSystVariationsSignal=False,
+    ):
+
+        if "charge" not in self.project:
+            self.project.append("charge")
+
+        # the number of systematics is not known before loading the histograms, the systematics are therefore stored in temporary dicts and sets
+        dict_noigroups = {}
+        dict_systgroups = {}
+
+        systsstandard = set()
+        systsnoconstraint = set()
+        systsnoprofile = set()
+
+        def book_systematic(syst, name):
+            if syst.get('noProfile', False):
+                systsnoprofile.add(name)
+            elif syst.get("noConstraint", False) or syst.get("noi", False):
+                systsnoconstraint.add(name)
+            else:
+                systsstandard.add(name)
+            group = syst["group"]
+            if group is None:
+                logger.warning(f"Systemtaic {name} is not a member of any group")
+                return 
+                #TODO: adding a group with the name of the member instead?
+                # group = name 
+            if syst.get("noi", False):
+                if group not in dict_noigroups:
+                    dict_noigroups[group] = set([name])
+                else:
+                    dict_noigroups[group].add(name)
+            else:
+                if group not in dict_systgroups:
+                    dict_systgroups[group] = set([name])
+                else:
+                    dict_systgroups[group].add(name)
+
+        #list of groups of systematics (nuisances) and lists of indexes
```

Comment:
This could really use some shorter helper functions to make it easier to follow/review

Reply:
Done